qt/widgets/instrumentview/src/Shape2DCollection.cpp

```diff
@@ -532,6 +533,43 @@ void Shape2DCollection::removeSelectedShapes() {
   }
 }
 
+/**
+ * @brief Shape2DCollection::copySelectedShapes
+ * Add the selected shapes to a copy buffer. Remove those previously stored.
+ */
+void Shape2DCollection::copySelectedShapes() {
```

Comment:
This should really be done in a presenter/model - I know one does not exist for this file, but it would be a good start to add one to start bringing us in-line with the GUI standards and making testing this not a PITA...as I am going to ask you to test it (sorry)

I would:
1. Create a presenter and a model, rename `copySelectedShapes` to `getSelectedShapes` and return a std::vector instead [to avoid leaking Qt Types](http://developer.mantidproject.org/Standards/GUIStandards.html#do-not-let-qt-types-in)

Then you just need:

Shape2DCollection.cpp (AKA View):
```
Signal / Slot:
   // Calls presenter instead of invoking directly the view

std::vector<std::unique_ptr<Shape2D>> getSelectedShapes(){
    std::vector<std::unique_ptr<Shape2D>> copiedShapes;
    auto newShape = shape->clone();
    newShape->setFillColor(shape->getFillColor());

    copiedShapes.pushback(std::unique_ptr<Shape2D>());
    return copiedShapes;
}

void setCopiedShapes(std::vector<std::unique_ptr<Shape2D>> shapes){
   for (auto shape: shapes){
     // Add a comment explaining why this second clone is needed so we don't take it out by accident
      auto pastedShape = shape->clone()
      addShape(pastedShape, false);
   }
```

Presenter.cpp:
```
void copySelectedShapes(){
    // Note we call the view - this makes testing easy
    auto shapes = view.getSelectedShapes();
    model.storeSelectedShapes(shapes);
}

void pasteCopiedShapes(){
    auto shapes = model.getCopiedShapes();
    view.setCopiedShapes(shapes)
```

Model.cpp
```
void storeCopiedShapes(std::vector<std::unique_ptr<Shape2D>> shapes){
    m_shapes = shapes;
}

std::vector<std::unique_ptr<Shape2D>  getCopiedShapes(){
    for (auto shape : m_shapes){

    if (shape->type() == "sector") {
      double angleOffset =
          shape->getDouble("endAngle") - shape->getDouble("startAngle");
      shape->setDouble("startAngle",
                       shape->getDouble("startAngle") + angleOffset);

      shape->setDouble("endAngle", shape->getDouble("endAngle") + angleOffset);
    } else {
      shape->moveBy(QPointF(0.1, 0.1));
    }
  }

  return shapes;
}

```

Then you can add tests - which is why we do this complex wiring:

View: The view can't be tested easily hence we made it as dumb as possible
Presenter: Use mocking to introduce a mock view and model. Check that `getSelectedShapes()` or `pasteCopiedShaped()` and the similar model methods are called as required 
Model: This should just be normal unit testing, you should check that the start / end angle set are what you expect

----

- We use unique_ptr to avoid leaking memory, since we clone (and take responsibility) when Qt no longer holds it. Release means Qt manages the lifetime again. This avoids a leak when a user copies twice without pasting the first time

Feel free to ask any questions, as this is quite a large change set